go-controller/pkg/ovn/egressfirewall_dns.go

```diff
@@ -0,0 +1,178 @@
+	// this map holds dnsNames to the dnsEntries
+	dnsEntries map[string]*dnsEntry
+
+	// Report change when Add operation is done
+	added    chan bool
+	stopChan chan struct{}
+}
+
+type dnsEntry struct {
+	//this map holds all the namespaces that a dnsName appears in
+	namespaces map[string]struct{}
+	// the current IP addresses the dnsName resolves to
+	dnsResolves []net.IP
+	// the addressSet that contains the current IPs
+	dnsAddressSet AddressSet
+}
+
+func NewEgressDNS() (*EgressDNS, error) {
+	dnsInfo, err := util.NewDNS("/etc/resolv.conf")
+	if err != nil {
+		utilruntime.HandleError(err)
+		return nil, err
+	}
+
+	egressDNS := &EgressDNS{
+		dns:        dnsInfo,
+		dnsEntries: make(map[string]*dnsEntry),
+
+		added:    make(chan bool),
+		stopChan: make(chan struct{}),
+	}
+
+	go utilwait.Until(egressDNS.Sync, 0, egressDNS.stopChan)
+
+	return egressDNS, nil
+}
+
+func (e *EgressDNS) Add(namespace, dnsName string, addressSet AddressSet) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	if _, exists := e.dnsEntries[dnsName]; !exists {
+		dnsEntry := dnsEntry{
+			namespaces: make(map[string]struct{}),
+		}
+		dnsEntry.namespaces[namespace] = struct{}{}
+		dnsEntry.dnsAddressSet = addressSet
+		e.dnsEntries[dnsName] = &dnsEntry
+		//only call Add if the dnsName doesn't exist in dnsEntries
+		if err := e.dns.Add(dnsName); err != nil {
+			utilruntime.HandleError(err)
+		}
+		e.signalAdded()
+	} else {
+		e.dnsEntries[dnsName].namespaces[namespace] = struct{}{}
+
+	}
+	e.dnsEntries[dnsName].dnsResolves = e.dns.GetIPs(dnsName)
+	err := addressSet.AddIPs(e.dnsEntries[dnsName].dnsResolves)
+	if err != nil {
+		klog.Errorf("Cannot add IPs to EgressFirewalls AddressSet: %s - %v", dnsName, err)
+	}
+
+}
+
+func (e *EgressDNS) Delete(namespace string) bool {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	// go through all dnsNames for namespaces
+	for dnsName, dnsEntry := range e.dnsEntries {
+		// delete the dnsEntry
+		delete(dnsEntry.namespaces, namespace)
+		if len(dnsEntry.namespaces) == 0 {
+			// the dnsEntry appears in no other namespace so delete the address_set
+			err := dnsEntry.dnsAddressSet.Destroy()
+			if err != nil {
+				klog.Errorf("Error deleteing EgressFirewall AddressSet for dnsName: %s %v", dnsName, err)
+			}
+			// the dnsEntry is no longer needed because nothing referances it delete it
+			delete(e.dnsEntries, dnsName)
+		}
+	}
+	return len(e.dnsEntries) == 0
+}
+
+func (e *EgressDNS) Update(dns string) (bool, error) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	return e.dns.Update(dns)
+}
+
+func (e *EgressDNS) Sync() {
+	var duration time.Duration
+	for {
+		tm, dnsName, ok := e.GetNextQueryTime()
+		if !ok {
+			duration = 30 * time.Minute
+		} else {
+			now := time.Now()
+			if tm.After(now) {
+				duration = tm.Sub(now)
+			} else {
+				_, err := e.Update(dnsName)
+				if err != nil {
+					utilruntime.HandleError(err)
+				}
+
+				if !reflect.DeepEqual(e.dnsEntries[dnsName].dnsResolves, e.dns.GetIPs(dnsName)) {
```

Comment:
This is order dependent; either use the `ipsEqual` function or sort the list and check for equality.

Reply:
sorted the list